app/src/main/java/org/openobservatory/ooniprobe/activity/MainActivity.java

```diff
@@ -129,6 +136,45 @@ public void onDrawerStateChanged(int newState){
         }
 
         checkInformedConsent();
+
+        // XXX: This is probably not correct: we would like to send
+        // info to the orchestrator only when the network or any other
+        // orchestrator parameter like country code changed.
+        String token = FirebaseInstanceId.getInstance().getToken();
+        if (token != null) {
+            NotificationService ns = NotificationService.getInstance(this);
+            ns.setDevice_token(token);
+            ns.sendRegistrationToServer();
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        final MainActivity activity = this;
+        IntentRouter.getInstance(getApplicationContext())
+            .register_handler("main_activity", "orchestrate/notification",
+                new IntentCallback() {
+                    @Override
+                    public void callback(Intent intent) {
+                        Log.d(TAG, intent.getStringExtra("message"));
+                        AlertDialog.Builder builder = new AlertDialog.Builder(
+                            activity
+                        );
+                        builder
+                            .setMessage(intent.getStringExtra("message"))
+                            .setTitle("ORCHESTRATION");
+                        AlertDialog dialog = builder.create();
+                        dialog.show();
```

Comment:
Note to self: I don't understand where we're telling the code that, when we click, it should open the browser.